### PR TITLE
Changed TOSA common legalizations to use llvm::Optional<>

### PR DESCRIPTION
**Pull request by jsmolens**
This is one of the TODOs from https://github.com/tensorflow/tensorflow/pull/44851

- Changed TOSA common legalizations to use llvm::Optional<> with Value
  ValueRange, as appopriate
- Removed legalizations return type macros
- Added an affirmative check for the "NONE" type in
  convertFusedActivation() to avoid an ambiguous meaning for the
  nullptr return value when the return type was an Operator *

### CODE REVIEW

File: tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc [hunk 1]
```diff
@@ -118,13 +106,17 @@ LogicalResult ConvertTFBiasAddOp::matchAndRewrite(
   // Bias tensor that feeds into tosa.conv2d must be rank 1
   if (bias_shape.size() != 1) return failure();
 
-  auto lowered_op = convertTFConv2DCommon(
+  auto result = convertTFConv2DCommon(
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:109]: Nit: can you spell out the type here?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 2]
```diff
@@ -72,7 +67,7 @@ Operation* convertPackOp(PatternRewriter& rewriter, Operation* op,
   // Check for ranked tensor type.
   if (!result_type) {
     op->emitOpError("PackOp: result type not ranked tensor");
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:69]: Please as followup, replace all the "emitOpError" with rewriter.notifyMatchFailure()`.

In particular in legalization the framework expects some pattern to fail to apply and this is not an error: other pattens may be able to handle it.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 3]
```diff
@@ -228,42 +223,38 @@ Operation* convertPackOp(PatternRewriter& rewriter, Operation* op,
...
+  auto reshape_output_type = RankedTensorType::get(
+      ArrayRef<int64_t>(reshape_output_shape), result_type.getElementType());
+
+  auto a2_reshape_op = rewriter.create<tosa::ReshapeOp>(
+      op->getLoc(), reshape_output_type, a1_concat_op.getResult(), shape_attr);
+
+  // If axis is equal to input tensor rank, then we need extra transpose
+  // [N, A, B, C] to [A, B, C, N]
+  if (axis == input_tensor_rank) {
+    auto a3_transpose_perm =
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:240]: Nit: spell out the type instead of auto

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc [hunk 4]
```diff
@@ -354,71 +345,73 @@ Operation* convertUnpackOp(PatternRewriter& rewriter, Operation* op,
...
-    auto reshape_op = rewriter.create<tosa::ReshapeOp>(
-        op->getLoc(),
-        RankedTensorType::get(ArrayRef<int64_t>(new_cond_dims),
-                              condition_type.getElementType()),
-        condition_value, rewriter.getI64ArrayAttr(new_cond_dims));
+  // Need to reshape the condition.
+  SmallVector<int64_t, 8> new_cond_dims;
+  for (int i = 0; i < (result_type.getRank() - condition_type.getRank()); i++) {
+    new_cond_dims.push_back(1);
+  }
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:382]: Isn't this equivalent to 
```
  SmallVector<int64_t, 8> new_cond_dims(result_type.getRank() - condition_type.getRank()), 1);
```
?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc [hunk 5]
```diff
@@ -351,39 +328,48 @@ LogicalResult ConvertTFSubOp::matchAndRewrite(Operation *op,
   return success();
 }
 
-LogicalResult ConvertTFMulOp::matchAndRewrite(Operation *op,
-                                              PatternRewriter &rewriter) const {
+LogicalResult ConvertTFMulOp::matchAndRewrite(Operation* op,
+                                              PatternRewriter& rewriter) const {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:332]: Please avoid formatting changes in unrelated part of the code, this makes the diff unreadable.

**Review by joker-eph — COMMENTED** [review]
You change from Operation* to Optional<Value> seem to assume a single result everywhere, There aren't any multi-results op?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc [hunk 6]
```diff
@@ -738,35 +739,43 @@ LogicalResult ConvertTFShapeOp::matchAndRewrite(
...
   auto tf_expanddims_op = cast<TF::ExpandDimsOp>(op);
 
-  auto lowered_op =
+  auto result =
       convertExpandDimsOp(rewriter, op, tf_expanddims_op.getResult(),
                           tf_expanddims_op.input(), tf_expanddims_op.dim());
 
-  TOSA_REPLACE_LOWERED_OP(rewriter, op, lowered_op);
+  if (!result) return failure();
+
```
Comment (stellaraccident) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:750]: Nit: through all of these: I think it is fine to collapse these blank lines. If you decide to do it, do it to all of them.

**Review by stellaraccident — COMMENTED** [review]
Modulo nits and changes requested by Mehdi, lgtm.

**Comment by stellaraccident** [comment]
Thanks for the cleanup!

**Comment by jsmolens** [comment]
> 
> 
> You change from Operation* to Optional seem to assume a single result everywhere, There aren't any multi-results op?

There are only a few multi-result ops that have common legalization functions (search for 'ValueRange').  One example in legalize_common.h:63

llvm::Optional<ValueRange> convertUnpackOp(PatternRewriter& rewriter,

File: tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc (hunk 1 above)
Comment (jsmolens) [reply tensorflow/compiler/mlir/tosa/transforms/fuse_bias_tf.cc:109]: Will do for the auto types here.  I'll try to spell out the other auto types, where possible.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc (hunk 4 above)
Comment (jsmolens) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:382]: Thanks for the hint!  I wasn't aware of this SmallVector constructor.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc (hunk 5 above)
Comment (jsmolens) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:332]: @joker-eph, @stellaraccident:

Ugh.   Looking back to the initial patch from Suraj which had similar formatting issues which were fixed for us and this patch, I think there is some disagreement between how our editors (emacs 25.2.2 on Ubuntu18.04 w/clang-6, I believe) are interpreting clang-format -style=Google and what the Google clang-format environment _actually_ is.   

This is the formatting that was chosen by the default clang-format picked up from tensorflow/.clang-format in our environment.  Interestingly, if I remove that file and put in my own clang-format -style=Google -dump-config, I get yet another interpretation. 

Obviously, this will be an irritating problem going forward if we don't get to the bottom of what's causing it.  Does this sort of issue ring any bells?

Comment (jsmolens) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:332]: I think I roughly figured out what's happening here.   Could I kindly ask for a pass on this, just this one time?  

TL;DR; the original clang-formatting does is wrong on legalize_tf.cc

The initial TOSA commit was formatted using -style=Google, which uses the clang-format parameter DerivePointerAlignment: true.  It should have used the default .clang-format in tensorflow/.clang-format, which sets that parameter to false.  DerivePointerAlignment is one of those mind-boggling parameters that chooses the setting for PointerAlignment based on the dominant usage already in the file.  For most of the files, it chose PointerAlignment: left, but for legalize_tf.cc, it chose PointerAlignment: right instead.    

The only thing the default TensorFlow clang-format file does is disable this per-file behavior and format PointerAlignment: left unconditionally.  

So we're left with a few choices:
1. Commit this file with some unrelated, but correct clang reformatting (what we have now, modulo the other todos)
2. Commit this file while emulating the incorrect clang reformatting with a manual override on clang-format (and possible validation failures later on?)
3. Insert and merge a separate commit before this one that corrects the issue, merge, rebase.

My preference is for (1) because it's simpler mechanically, although it does have the ugly reformatting + code changes.
We should definitely scrub the files that weren't touched in this commit to see if there are any other affected files.

Comment (stellaraccident) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:332]: I'm fine with the first option. Dueling tools are not worth the terrible to diagnose in flight at this level. When you're ready, I can commit this and apply any formatting fixups as part of that.

Comment (joker-eph) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:332]: In general we use *git clang-format* and never format the entire file (so don't use editor full-file formatting): so it does not matter much that you have some local discrepancies with the clang-format version. The point is more about not exploding the size of the diff.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc [hunk 7]
```diff
@@ -2003,7 +2019,7 @@ LogicalResult ConvertTFFakeQuantWithMinMaxVarsOp::matchAndRewrite(
     Operation* op, PatternRewriter& rewriter) const {
   auto tf_fakequant_op = cast<TF::FakeQuantWithMinMaxVarsOp>(op);
 
-  auto output_type =
+  RankedTensorType output_type =
       tf_fakequant_op.getResult().getType().dyn_cast<RankedTensorType>();
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:2023]: For future reference: `auto` is perfectly fine when the type is obvious from the context, like when it appears in the "cast" template parameter on the RHS like here.

In general the guideline is to use `auto` when it makes the code more readable though (like iterator and other hard to write types).

**Review by joker-eph — APPROVED** [review]
